### PR TITLE
capture per monitor if they are available

### PR DESCRIPTION
**Pull request by jsorg71**

### CODE REVIEW

File: module/rdpCapture.c [hunk 1]
```diff
@@ -756,12 +755,12 @@ rdpCapture1(rdpClientCon *clientCon,
         if (ex != 0)
         {
             rect.x2 += ex;
-            if (rect.x2 > min_width)
+            if (rect.x2 > srect.x2)
             {
-                rect.x1 -= rect.x2 - min_width;
-                rect.x2 = min_width;
+                rect.x1 -= rect.x2 - srect.x2;
```
Comment (proski) [inline module/rdpCapture.c]: I feel this part became less clear and needs a comment.

File: module/rdpCapture.c [hunk 2]
```diff
@@ -887,11 +860,13 @@ rdpCapture2(rdpClientCon *clientCon,
...
-    rect.x1 = 0;
-    rect.y1 = 0;
-    rect.x2 = min(dst_width, src_width);
-    rect.y2 = min(dst_height, src_height);
+    rect.x1 = src_left;
+    rect.y1 = src_top;
+    rect.x2 = src_left + src_width;
+    rect.x2 = RDPMIN(dst_width, rect.x2);
+    rect.y2 = src_top + src_height;
+    rect.y2 = RDPMIN(dst_height, rect.y2);
```
Comment (proski) [inline module/rdpCapture.c]: Can this part be shared between all rdpCaptureN functions?

File: module/rdpClientCon.c [hunk 3]
```diff
@@ -2170,14 +2170,23 @@ rdpDeferredUpdateCallback(OsTimerPtr timer, CARD32 now, pointer arg)
     int num_rects;
     struct image_data id;
 
+    int index;
```
Comment (proski) [inline module/rdpClientCon.c]: Not a good variable name.

File: module/rdpClientCon.c [hunk 4]
```diff
@@ -2195,30 +2204,114 @@ rdpDeferredUpdateCallback(OsTimerPtr timer, CARD32 now, pointer arg)
...
-                   clientCon->cap_width, clientCon->cap_height,
-                   clientCon->cap_stride_bytes,
-                   clientCon->rdp_format, clientCon->client_info.capture_code))
-    {
-        LLOGLN(10, ("rdpDeferredUpdateCallback: num_rects %d", num_rects));
-        rdpClientConSendPaintRectShmEx(clientCon->dev, clientCon, &id,
-                                       clientCon->dirtyRegion,
-                                       rects, num_rects);
-        free(rects);
+    if (clientCon->dev->monitorCount < 1)
```
Comment (proski) [inline module/rdpClientCon.c]: Does the non-multimon need to be handled specially? I would prefer to share as much code as possible. I see quite a lot of similar code.

Comment (proski) [inline module/rdpClientCon.c]: Here's a design pattern I learned from complex projects. Always code for the hardest case. That gives good input for design. Then port to easier cases. Keep additional logic outside the original code by using short functions and macros.

**Comment by proski** [comment]
Why is it good to capture per monitor? How would it benefit users?

**Comment by jsorg71** [comment]
This is what Windows 7.1 RemoteFX does.  I think you don't want to do the whole desktop because less memory and more responsive with smaller tile sets.  Also, don't break up smaller than a monitor because then you see tearing.

**Comment by proski** [comment]
OK, makes sense.

File: module/rdpClientCon.c [hunk 5]
```diff
@@ -2195,30 +2247,92 @@ rdpDeferredUpdateCallback(OsTimerPtr timer, CARD32 now, pointer arg)
...
+            if (band_index == band_count)
+            {
+                /* gone through all bands, nothing changed */
+                rdpRegionDestroy(clientCon->dirtyRegion);
+                clientCon->dirtyRegion = rdpRegionCreate(NullBox, 0);
+            }
+        }
+        else
+        {
+            /* nothing changed in visable area */
```
Comment (proski) [inline module/rdpClientCon.c]: visable -> visible

File: module/rdpClientCon.c [hunk 6]
```diff
@@ -2162,13 +2162,65 @@ rdpClientConSendPaintRectShmEx(rdpPtr dev, rdpClientCon *clientCon,
 }
 
 /******************************************************************************/
+static int
+rdpCapRect(rdpClientCon *clientCon, BoxPtr cap_rect, struct image_data *id)
```
Comment (proski) [inline module/rdpClientCon.c]: Would it be possible to add comments to new functions? The name is too short to be clear.

File: module/rdpCapture.c [hunk 7]
```diff
@@ -632,72 +604,70 @@ rdpCapture0(rdpClientCon *clientCon,
         (*out_rects)[i] = rect;
     }
 
-    if ((src_format == XRDP_a8r8g8b8) && (dst_format == XRDP_a8r8g8b8))
+    src = id->pixels;
+    dst = id->shmem_pixels;
+    dst_format = clientCon->rdp_format;
+    src_stride = id->lineBytes;
+    dst_stride = clientCon->cap_stride_bytes;
```
Comment (jsorg71) [inline module/rdpCapture.c]: That depends on the version of xorg you are building against and the system you are building on.  Xorg used to be way behind on there compile requirement.  But you're right, the latest one is using C99 stuff.  I think sprinkling variables allover functions is another issue.

File: module/rdpClientCon.c (hunk 6 above)
Comment (jsorg71) [inline module/rdpClientCon.c]: Sure, I was thinking that when I added it.

**Comment by jsorg71** [comment]
I'm going to merge this one as people are waiting on it.
We can do a new PR for other items.

**Pull request merged**